### ISSUE TITLE
Config Server with BitBucket git repo on Google App Engine failing several times per day with lock error

### ISSUE THREAD

**Issue by bjornharvold** [spring-cloud/spring-cloud-config#726]
I am running Spring Boot 1.5.4 for all my apps.

I have a config server servicing several other spring boot apps on GAE. It keeps on going down several times per day with the following message:

```
A  org.eclipse.jgit.errors.LockFailedException: Cannot lock /~/.git/index
	at org.eclipse.jgit.dircache.DirCache.lock(DirCache.java:250) ~[org.eclipse.jgit-4.6.0.201612231935-r.jar!/:4.6.0.201612231935-r]
	at org.eclipse.jgit.dircache.DirCache.lock(DirCache.java:327) ~[org.eclipse.jgit-4.6.0.201612231935-r.jar!/:4.6.0.201612231935-r]
	at org.eclipse.jgit.dircache.DirCache.lock(DirCache.java:293) ~[org.eclipse.jgit-4.6.0.201612231935-r.jar!/:4.6.0.201612231935-r]
	at org.eclipse.jgit.lib.Repository.lockDirCache(Repository.java:1192) ~[org.eclipse.jgit-4.6.0.201612231935-r.jar!/:4.6.0.201612231935-r]
	at org.eclipse.jgit.api.CheckoutCommand.call(CheckoutCommand.java:265) ~[org.eclipse.jgit-4.6.0.201612231935-r.jar!/:4.6.0.201612231935-r]
	at org.springframework.cloud.config.server.environment.JGitEnvironmentRepository.checkout(JGitEnvironmentRepository.java:263) ~[spring-cloud-config-server-1.3.1.RELEASE.jar!/:1.3.1.RELEASE]
	at org.springframework.cloud.config.server.environment.JGitEnvironmentRepository.refresh(JGitEnvironmentRepository.java:195) ~[spring-cloud-config-server-1.3.1.RELEASE.jar!/:1.3.1.RELEASE]
	at org.springframework.cloud.config.server.environment.JGitEnvironmentRepository.getLocations(JGitEnvironmentRepository.java:169) ~[spring-cloud-config-server-1.3.1.RELEASE.jar!/:1.3.1.RELEASE]
	at org.springframework.cloud.config.server.environment.MultipleJGitEnvironmentRepository.getLocations(MultipleJGitEnvironmentRepository.java:146) ~[spring-cloud-config-server-1.3.1.RELEASE.jar!/:1.3.1.RELEASE]
	at org.springframework.cloud.config.server.environment.AbstractScmEnvironmentRepository.findOne(AbstractScmEnvironmentRepository.java:42) ~[spring-cloud-config-server-1.3.1.RELEASE.jar!/:1.3.1.RELEASE]
	at org.springframework.cloud.config.server.environment.MultipleJGitEnvironmentRepository.findOne(MultipleJGitEnvironmentRepository.java:186) ~[spring-cloud-config-server-1.3.1.RELEASE.jar!/:1.3.1.RELEASE]
	at org.springframework.cloud.config.server.environment.CompositeEnvironmentRepository.findOne(CompositeEnvironmentRepository.java:45) ~[spring-cloud-config-server-1.3.1.RELEASE.jar!/:1.3.1.RELEASE]
	at org.springframework.cloud.config.server.environment.EnvironmentEncryptorEnvironmentRepository.findOne(EnvironmentEncryptorEnvironmentRepository.java:53) ~[spring-cloud-config-server-1.3.1.RELEASE.jar!/:1.3.1.RELEASE]
	at org.springframework.cloud.config.server.environment.EnvironmentController.labelled(EnvironmentController.java:106) ~[spring-cloud-config-server-1.3.1.RELEASE.jar!/:1.3.1.RELEASE]
	at org.springframework.cloud.config.server.environment.EnvironmentController.defaultLabel(EnvironmentController.java:95) ~[spring-cloud-config-server-1.3.1.RELEASE.jar!/:1.3.1.RELEASE]
	at sun.reflect.GeneratedMethodAccessor84.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_131]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_131]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205) ~[spring-web-4.3.9.RELEASE.jar!/:4.3.9.RELEASE]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:133) ~[spring-web-4.3.9.RELEASE.jar!/:4.3.9.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:97) ~[spring-webmvc-4.3.9.RELEASE.jar!/:4.3.9.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:827) ~[spring-webmvc-4.3.9.RELEASE.jar!/:4.3.9.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:738) ~[spring-webmvc-4.3.9.RELEASE.jar!/:4.3.9.RELEASE]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85) ~[spring-webmvc-4.3.9.RELEASE.jar!/:4.3.9.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967) ~[spring-webmvc-4.3.9.RELEASE.jar!/:4.3.9.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901) ~[spring-webmvc-4.3.9.RELEASE.jar!/:4.3.9.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970) ~[spring-webmvc-4.3.9.RELEASE.jar!/:4.3.9.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861) ~[spring-webmvc-4.3.9.RELEASE.jar!/:4.3.9.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:635) ~[tomcat-embed-core-8.5.15.jar!/:8.5.15]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846) ~[spring-webmvc-4.3.9.RELEASE.jar!/:4.3.9.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:742) ~[tomcat-embed-core-8.5.15.jar!/:8.5.15]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) ~[tomcat-embed-core-8.5.15.jar!/:8.5.15]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-8.5.15.jar!/:8.5.15]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) ~[tomcat-embed-websocket-8.5.15.jar!/:8.5.15]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-8.5.15.jar!/:8.5.15]
```

Looks like someone reported something similar here:
https://www.eclipse.org/forums/index.php/t/1075875/

The errors start can start showing up in the logs immediately when starting config server or it takes a few hours and sometimes even days. Not sure what to make of this.

**Comment by bjornharvold** [comment]
To follow up with some extra information. Config Server instance can potentially scale to more than 1. I also have 4 environments DEV, QA, UAT, PROD; each running an instance of config server talking to the same joint BitBucket repo but using different encryption keys.

**Comment by dsyer** [comment]
The Spring Boot version is relevant but not that informative. What about the Spring Cloud version? How did you configure the git repositories?

**Comment by bjornharvold** [comment]
Thank you for getting back to me so quickly.

Here's the application.yml for config server:
```
logging:
  level:
    ROOT: INFO

spring:
  application:
    name: traveliko-web-config-server
  cloud:
    config:
      server:
        git:
          uri: https://bitbucket.org/project/mygitrepo.git
          username: username
          password: password
          force-pull: true
          basedir: ~/
      label: master

security:
  user:
    name: @config.server.username@
    password: @config.server.password@
```
We are using Dalston.SR1.

On a side note: We think we are seeing a side-effect of this as well. When config server goes down, the spring boot apps that are dependent on config server sometimes fails with something akin to this error:
https://jira.spring.io/browse/SPR-12902

GAE tries to recover by reloading the app and the app goes down because config server is down.

So definitely interested in getting to the bottom of this.

**Comment by dsyer** [comment]
You didn't say what version of Spring Cloud? Any other ideas how to reproduce the issue?

**Comment by bjornharvold** [comment]
<dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-config-server</artifactId>
        </dependency>

<dependency>
                <groupId>org.springframework.cloud</groupId>
                <artifactId>spring-cloud-dependencies</artifactId>
                <version>Dalston.SR1</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>

To reproduce in a nutshell:
- Create a config server spring boot app
- Connect to a bitbucket repo
- Deploy on Google App Engine

There is no changes being made to the config files on BitBucket for this error to occur. I am not sure whether querying the config server from another app makes a difference or not. It looks like there is a timer in config server that checks for changes with the Bitbucket repo and it is there the error occurs.

We are using a flexible GAE application so we have some ephemeral disk space to play with. Writing to the VM file system 'should" not be the issue.

Let me know if you have any other questions. Keen on fixing this.

**Comment by dsyer** [comment]
"Deploy on Google App Engine" is a bit of a dark secret to me, and it looks like the only unusual thing in your recipe. Could you create a sample and put a README in to help someone to do this accurately?

**Comment by bjornharvold** [comment]
Will shoot off a sample shortly

**Comment by dsyer** [comment]
It looks to me like the only way this could happen is if you have 2 instances of the config server sharing a working copy (i.e. the same filesystem). Since I have no way to know how your instances on GAE share state, I'll have to leave it up to you to investigate that.

N.B. `basedir: ~/` isn't doing what you want it to (it creates a directory called "~", as can be seen from the error message when the lock fails). Maybe a random value would be better anyway, e.g. `basedir=${user.home}/${random.value}`?

**Comment by bjornharvold** [comment]
[demo.zip](https://github.com/spring-cloud/spring-cloud-config/files/1097182/demo.zip)

Here's a sample with readme. 

- I have set the scaling to 1 a while back but with the same issue
- I need to look in the GAE docs to see where that ~ comes from. I don't know if I have control over that. Will report back.

**Comment by bjornharvold** [comment]
Apologies, I forgot to enable Config Server in my sample:

[demo.zip](https://github.com/spring-cloud/spring-cloud-config/files/1097188/demo.zip)

**Comment by dsyer** [comment]
Thanks for the README. I don't actually believe for a minute that this will fail in the way you describe. If there are no clients connecting, then there will be no activity on the server at all, so it cannot produce the stack trace you posted. Maybe if you could work out a way to actually cause it to fail, then we could make progress? Or set up a server using an account on GCP that you can share with me (I don't know how they do that, e.g. on Amazon you could add me to an IAM group). My google ID is david.sayer@googlemail.com.

**Comment by bjornharvold** [comment]
I have sent you access details. Take a look in logs under Production for config-server service. All the other apps there depend on config server running.

Thank you

**Comment by bjornharvold** [comment]
If this is not a Spring issue, please do let me know so I don't waste your time.

**Comment by dsyer** [comment]
Sorry, I mistyped my google ID, so I don't think I'm going to get that invitation.

**Comment by bjornharvold** [comment]
Shot off another invite with correct last name

**Comment by dsyer** [comment]
Please join https://gitter.im/spring-cloud/spring-cloud to follow up while we get the sample working (still waiting for an invitation here, and also waiting for GCP to finish uploading your sample to my account).

**Comment by bjornharvold** [comment]
It keeps on locking up even after baseUrl was changed. Same error as this issue as well:
https://github.com/spring-cloud/spring-cloud-config/issues/261

Going to see if there is a way to attach some non-ephemeral storage to my GAE app.

**Comment by bjornharvold** [comment]
No such thing unfortunately. Moving my config server instance over to EC2.

**Comment by dsyer** [comment]
It's still a bit mysterious, but I did reproduce the problem with my own GAE app. My best theory so far for what happens is that the server comes under load and GAE kills it when it exceeds a threshold for slow responses, but it doesn't clean up the storage, so there is a stale lock file left over when it restarts.

This would fix it: e6a42aab9.  Probably worth trying a snaphot if you get a chance.
